### PR TITLE
Add auth

### PR DESCRIPTION
**Pull request by ryanlyrl**

### CODE REVIEW

File: server/src/main/kotlin/com/cuhacking/watershed/resources/AuthResource.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import javax.inject.Inject
+
+@Serializable
+data class InputCredentials(val email: String, val password: String)
+
+class AuthResource @Inject constructor(private val database: Database,
+                                       private val dispatcher: CoroutineDispatcher,
+                                       private val jwtManager: JwtManager) {
+    fun Routing.routing() {
+        route("/login") {
```
Comment (dellisd) [inline server/src/main/kotlin/com/cuhacking/watershed/resources/AuthResource.kt:24]: Aren't these `route` calls redundant? Could we just call `post("/login")` instead? We could reserve `route` calls for instances where an endpoint has multiple methods associated with it.

**Review by deenhaque — APPROVED** [review]
I 100% know what I just looked at.

LGTM 👍 🚀 🥇 🌔

File: server/src/main/kotlin/com/cuhacking/watershed/resources/AuthResource.kt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+                    call.respond(HttpStatusCode.OK, jwtManager.createJWT(user.uuid))
+                }
+            }
+        }
+
+        route("/private") {
+            authenticate("auth-jwt") {
+                get("/") {
+                    call.respond("Authenticated!")
```
Comment (dellisd) [inline server/src/main/kotlin/com/cuhacking/watershed/resources/AuthResource.kt:42]: Is this just for testing?

File: server/src/main/kotlin/com/cuhacking/watershed/server.kt [hunk 3]
```diff
@@ -36,20 +53,32 @@ class Main(configPath: String) {
...
+fun Application.installFeatures(config: Config, jwtManager: JwtManager) {
     install(CallLogging) {
         level = Level.DEBUG
     }
     install(AutoHeadResponse)
     install(ContentNegotiation) {
         json()
     }
+    install(Authentication) {
+        jwt("auth-jwt") {
```
Comment (dellisd) [inline server/src/main/kotlin/com/cuhacking/watershed/server.kt:65]: We could consider putting this string value as a constant int `JwtManager`'s companion object.

File: server/src/main/kotlin/com/cuhacking/watershed/util/JwtManager.kt [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import com.cuhacking.watershed.config.Config
+import com.cuhacking.watershed.db.Database
+import io.ktor.auth.jwt.*
+import java.util.*
+import javax.inject.Inject
+
+class JwtManager @Inject constructor(private val config: Config, private val database: Database){
+
+    val algorithm: Algorithm = Algorithm.HMAC512(config.auth.secret)
+    private val TOKEN_EXPIRY = 60*60*1000 // 1 hour
```
Comment (dellisd) [inline server/src/main/kotlin/com/cuhacking/watershed/util/JwtManager.kt:14]:
```suggestion
    private val TOKEN_EXPIRY = Duration.hour(1)
```

File: server/src/main/kotlin/com/cuhacking/watershed/util/JwtManager.kt [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+    fun createJWT(uuid: String) = JWT.create()
+        .withSubject("Authentication")
+        .withIssuer(config.auth.issuer)
+        .withClaim("uuid", uuid)
+        .withExpiresAt(getExpiration())
+        .sign(algorithm)
+
+    private fun getExpiration() = Date(System.currentTimeMillis() + TOKEN_EXPIRY)
```
Comment (dellisd) [inline server/src/main/kotlin/com/cuhacking/watershed/util/JwtManager.kt:23]:
```suggestion
    private fun getExpiration() = Date(System.currentTimeMillis() + TOKEN_EXPIRY.toInt(DurationUnit.Milliseconds)
```
Could also add an extension function `Duration.toMilliseconds()` if that's too much to write lol

**Review by dellisd — CHANGES_REQUESTED** [review]

File: server/src/main/kotlin/com/cuhacking/watershed/util/JwtManager.kt [hunk 6]
```diff
@@ -5,13 +5,18 @@ import com.auth0.jwt.algorithms.Algorithm
...
 import com.cuhacking.watershed.db.Database
 import io.ktor.auth.jwt.*
+import java.time.Duration
 import java.util.*
 import javax.inject.Inject
 
 class JwtManager @Inject constructor(private val config: Config, private val database: Database){
 
+    companion object {
+        val jwtAuth = "jwt-auth"
```
Comment (dellisd) [inline server/src/main/kotlin/com/cuhacking/watershed/util/JwtManager.kt:15]:
```suggestion
        val JWT_AUTH= "jwt-auth"
```

**Review by dellisd — APPROVED** [review]

**Pull request merged**